packages/editor/src/components/local-autosave-monitor/index.js

```diff
@@ -0,0 +1,169 @@
+		getEditedPostAttribute: select( 'core/editor' ).getEditedPostAttribute,
+	} ) );
+
+	return useCallback( () => {
+		window.localStorage.setItem( 'post_' + postId, JSON.stringify( {
+			post_title: getEditedPostAttribute( 'title' ),
+			content: getEditedPostAttribute( 'content' ),
+			excerpt: getEditedPostAttribute( 'excerpt' ),
+		} ) );
+	}, [ postId, getEditedPostAttribute ] );
+}
+
+/**
+ * Custom hook which manages the creation of a notice prompting the user to
+ * restore a local autosave, if one exists.
+ */
+function useAutosaveNotice() {
+	const {
+		postId,
+		getEditedPostAttribute,
+	} = useSelect( ( select ) => ( {
+		postId: select( 'core/editor' ).getCurrentPostId(),
+		getEditedPostAttribute: select( 'core/editor' ).getEditedPostAttribute,
+	} ) );
+
+	const { createWarningNotice, removeNotice } = useDispatch( 'core/notices' );
+	const { editPost, resetEditorBlocks } = useDispatch( 'core/editor' );
+
+	useEffect( () => {
+		let autosave = window.localStorage.getItem( 'post_' + postId );
+		if ( ! autosave ) {
+			return;
+		}
+
+		try {
+			autosave = JSON.parse( autosave );
+		} catch ( error ) {
+			// Not usable if it can't be parsed.
+			return;
+		}
+
+		const { post_title: title, content, excerpt } = autosave;
+		const edits = { title, content, excerpt };
+
+		// Only display a notice if there is a difference between what has been
+		// saved and that which is stored in localStorage.
+		const hasDifference = Object.keys( edits ).some( ( key ) => {
+			return edits[ key ] !== getEditedPostAttribute( key );
+		} );
+
+		if ( ! hasDifference ) {
+			// If there is no difference, it can be safely ejected from storage.
+			window.localStorage.removeItem( 'post_' + postId );
+
+			return;
+		}
+
+		const noticeId = uniqueId( 'wpEditorAutosaveRestore' );
+
+		createWarningNotice( __( 'The backup of this post in your browser is different from the version below.' ), {
+			id: noticeId,
+			actions: [
+				{
+					label: __( 'Restore the backup' ),
+					onClick() {
+						editPost( omit( edits, [ 'content' ] ) );
+						resetEditorBlocks( parse( edits.content ) );
+						removeNotice( noticeId );
+					},
+				},
+			],
+		} );
+	}, [ postId ] );
+}
+
+/**
+ * Custom hook which ejects a local autosave after a successful save occurs.
+ */
+function useAutosavePurge() {
+	const {
+		postId,
+		didSave,
+	} = useSelect( ( select ) => ( {
+		postId: select( 'core/editor' ).getCurrentPostId(),
+		didSave: select( 'core/editor' ).didPostSaveRequestSucceed(),
+	} ) );
+
+	const lastDidSave = useRef( didSave );
+
+	useEffect( () => {
+		if ( lastDidSave.current !== true && didSave ) {
+			window.localStorage.removeItem( 'post_' + postId );
+		}
+
+		lastDidSave.current = didSave;
+	}, [ didSave ] );
+}
+
+function LocalAutosaveMonitor() {
+	const autosave = useAutosaveSaveCallback();
+	useAutosaveNotice();
+	useAutosavePurge();
+
+	return (
+		<AutosaveMonitor
+			interval={ AUTOSAVE_INTERVAL_SECONDS }
```

Comment:
This seems to require that the user is idle for 15 seconds before it'll trigger a save.

Since this is a last resort, I wonder if there's also an option to use additional strategies for saving local autosaves, like the beforeUnload handler.

Reply:
Given that `localStorage` is a synchronous API, should [`requestIdleCallback`](https://developer.mozilla.org/en-US/docs/Web/API/Window/requestIdleCallback) be used if available.

Reply:
Perhaps related: https://github.com/WordPress/gutenberg/pull/16490/files#r303243055

Maybe we should only save locally if you cannot save otherwise? Would it make more sense then to implement it in a different layer of the application?

Reply:
> Since this is a last resort, I wonder if there's also an option to use additional strategies for saving local autosaves, like the beforeUnload handler.

I'd agree on this, and the fact that we should probably just schedule (or at least check for changes to justify) the autosaves to occur _every_ 15 seconds, not just 15 seconds after the last change.

> Given that `localStorage` is a synchronous API, should [`requestIdleCallback`](https://developer.mozilla.org/en-US/docs/Web/API/Window/requestIdleCallback) be used if available?

It could be worth exploring. I'd question whether it would make a measurable difference, however (related: #16657).